x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -1780,4 +1790,177 @@ private static LogicalPlan planWithoutSyntheticAttributes(LogicalPlan plan) {
+            // This rule should be applied after ResolveUnionTypes, so that the InvalidMappedFields with explicit casting are converted into
+            // MultiTypeEsField, and don't be double cast here.
+            Map<FieldAttribute, Alias> invalidMappedFieldCasted = new HashMap<>();
+            LogicalPlan transformedPlan = plan.transformUp(LogicalPlan.class, p -> {
+                if (p instanceof UnaryPlan == false && p instanceof LookupJoin == false) {
+                    return p;
+                }
+                Set<FieldAttribute> invalidMappedFields = invalidMappedFieldsInLogicalPlan(p);
+                if (invalidMappedFields.isEmpty() == false) {
+                    // If we are at a plan node that has invalid mapped fields, we need to either add an EVAL, or if that has been done
+                    // we should instead replace with the already cast field
+                    Map<FieldAttribute, Alias> newAliases = Maps.newHashMapWithExpectedSize(invalidMappedFields.size());
+                    Map<FieldAttribute, Alias> existingAliases = Maps.newHashMapWithExpectedSize(invalidMappedFields.size());
+                    for (FieldAttribute fa : invalidMappedFields) {
+                        if (invalidMappedFieldCasted.containsKey(fa)) {
+                            // There is already an eval plan created for the implicit cast field, just reference to it
+                            Alias alias = invalidMappedFieldCasted.get(fa);
+                            existingAliases.put(fa, alias);
+                        } else {
+                            // Create a new alias and later on add a new EVAL with this new aliases for implicit casting
+                            DataType targetType = commonDataType(fa);
+                            if (targetType != null) {
+                                Expression conversionFunction = castInvalidMappedField(targetType, fa);
+                                Alias alias = new Alias(fa.source(), fa.name(), conversionFunction);
+                                newAliases.put(fa, alias);
+                                invalidMappedFieldCasted.put(fa, alias);
+                            }
+                        }
+                    }
+                    // If there are new aliases created, create a new eval child with new aliases for the current plan。
+                    // How many children does a LogicalPlan have? Only deal with UnaryPlan and LookupJoin for now.
+                    if (newAliases.isEmpty() == false) { // create a new eval child plan
+                        if (p instanceof UnaryPlan u) { // unary plan
+                            Eval eval = new Eval(u.source(), u.child(), newAliases.values().stream().toList());
+                            p = u.replaceChild(eval);
+                        }
+                        // TODO Lookup join does not work on date_nanos field today, joining on a date_nanos field does not find a match.
```

Comment:
I've added a bug for date_nanos in LOOKUP JOIN at https://github.com/elastic/elasticsearch/issues/127249, and mentioned that this partial work on implicit casting needs to be done after (or during) that bugfix.